ntoskrnl/fsrtl/fastio.c

```diff
@@ -1837,113 +1869,61 @@ FsRtlAcquireFileForModWriteEx(IN PFILE_OBJECT FileObject,
         }
     }
 
-    Status = STATUS_SUCCESS;
-
-    /* No FastIo handler, use algorithm from Nagar p.550. */
-    if (!FcbHeader->Resource)
-    {
-        *ResourceToRelease = NULL;
-        return STATUS_SUCCESS;
-    }
+    /* Check what and how we should acquire */
+    Exclusive = FsRtlpGetResourceForModWrite(FcbHeader, EndingOffset, &ResourceToAcquire);
 
-    /* Default condition - shared acquiring of Paging IO Resource */
-    ResourceToAcquire = FcbHeader->PagingIoResource;
-
-    /* Decide on type of locking and type of resource based on historical magic
-       well explain by Nagar in p. 550-551 */
-    if ((EndingOffset->QuadPart > FcbHeader->ValidDataLength.QuadPart &&
-         FcbHeader->FileSize.QuadPart != FcbHeader->ValidDataLength.QuadPart) ||
-         (FcbHeader->Flags & FSRTL_FLAG_ACQUIRE_MAIN_RSRC_EX))
-    {
-        /* Either exclusive flag is set or write operation is extending
-           the valid data length. Prefer exclusive acquire then */
-        Exclusive = TRUE;
-        ResourceToAcquire = FcbHeader->Resource;
-    }
-    else if (!FcbHeader->PagingIoResource ||
-             (FcbHeader->Flags & FSRTL_FLAG_ACQUIRE_MAIN_RSRC_SH))
+    /* Acquire the resource and loop until we're sure we got this right. */
+    while(TRUE)
     {
-        /* Acquire main resource shared if flag is specified or
-           if PagingIo resource is missing */
-        Exclusive = FALSE;
-        ResourceToAcquire = FcbHeader->Resource;
-    }
+        BOOLEAN OldExclusive;
+        PERESOURCE OldResourceToAcquire;
 
-    /* Acquire the resource in the loop, since the above code is unsafe */
-    while (TRUE)
-    {
-        Result = FALSE;
+        if (ResourceToAcquire == NULL)
+        {
+            /* There's nothing to acquire. Success, I guess. */
```

Comment:
You guess, so you're not sure? Is it possible to get a test to determine that?

Reply:
I'm 90% sure there's no FS driver out there which would lead to such situation. I added an assertion there.

Reply:
Well, turns out that fastfat doesn't set any resource on its VPB header. Let's see how that goes without the ASSERT.